### ISSUE TITLE
Inconsisteny in as_numeric with indexed vars and expressions

### ISSUE THREAD

**Issue by gseastream** [Pyomo/pyomo#649]
I discovered an inconsistency between `IndexedVar` and `IndexedExpression` that is highlighted in the `as_numeric` function. IndexedVar has an `is_expression_type` method (which it inherits from `Var`), but IndexedExpression does not have this method. This comes up in as_numeric since in that function, it first checks if it is a native type (which neither of these are) and then it checks for the existence of the `is_expression_type` method. If it has this method, it returns the object that was passed to the function, and if it does not have this method, it will try another thing (which fails) and then throw an error and say this was an unexpected type. Thus, `as_numeric(IndexedVar)` will return the original IndexedVar, but `as_numeric(IndexedExpression)` will throw an error.

However, despite `IndexedVar` having the `is_expression_type` method, that is the only `NumericValue` is-type method it has. For example, it doesn't even have `is_variable_type`.

Was this inconsistency created deliberately? Should IndexedVar even have that method? I would think `as_numeric` should always return things that can be put into Pyomo expressions, but IndexedVars cannot be put into expressions, so perhaps we should remove that method (unless it is being used somewhere).

**Comment by jsiirola** [comment]
`as_numeric` is supposed to either return a valid `NumericValue`, or a valid object derived from / duck typing `NumericValue`.  IndexedVar is neither and really should raise an exception.  I am not sure why IndexedVar implements `is_expression_type` (`_VarData` should, and by extension `SimpleVar`, but not `IndexedVar`).

This is really a question for @whart222.

In the interim, what happens if you remove `is_expression_type` from `IndexedVar`?

**Comment by gseastream** [comment]
`Var` implements `is_expression_type`, and `IndexedVar` inherits that. `IndexedVar` does not implement it itself. That being said obviously that's still a problem.

If you remove the method and run the `pyomo.core` tests, the following tests raise the following error:

```
pyomo.core.tests.unit.test_expr_pyomo5.TestGeneralExpressionGeneration.test_invalidIndexing
pyomo.core.tests.unit.test_expr_pyomo5.TestGenerate_ChainedRelationalExpression.test_inequalityErrors
pyomo.core.tests.unit.test_expr_pyomo5.TestGenerate_RelationalExpression.test_equalityErrors

AttributeError: 'IndexedVar' object has no attribute 'is_expression_type'
```

**Comment by ghackebeil** [comment]
If I recall correctly, it was thrown on to make it possible to provide a helpful error message when a user attempts to use an indexed variable in an expression (without providing an index).

**Comment by blnicho** [comment]
I think this was fixed in #1433 so closing.
